### PR TITLE
add support device assurance

### PR DESCRIPTION
**Pull request by duytiennguyen-okta**

### CODE REVIEW

**Comment by lucascantor** [comment]
@duytiennguyen-okta thank you so much for working on this. It sincerely means the world to me.

File: okta/resource_okta_policy_device_assurance_androi_test.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+`
+)
+
+var (
+	testAccProtoV6ProviderFactories = map[string]func() (tfprotov6.ProviderServer, error){
+		"okta": providerserver.NewProtocol6WithError(NewFWProvider("test")),
+	}
+)
+
+func TestAccPolicyDeviceAssuranceAndroid(t *testing.T) {
```
Comment (monde) [inline okta/resource_okta_policy_device_assurance_androi_test.go]: TestAccResourceOktaPolicyDeviceAssuranceAndroid

File: okta/resource_okta_policy_device_assurance_chromeOS_test.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
+package okta
+
+import (
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func TestAccPolicyDeviceAssuranceChromeOS(t *testing.T) {
```
Comment (monde) [inline okta/resource_okta_policy_device_assurance_chromeOS_test.go]: TestAccResourceOktaPolicyDeviceAssuranceChromeOS

File: okta/resource_okta_policy_device_assurance_iOS_test.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
+package okta
+
+import (
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func TestAccPolicyDeviceAssuranceIOS(t *testing.T) {
```
Comment (monde) [inline okta/resource_okta_policy_device_assurance_iOS_test.go]: TestAccResourceOktaPolicyDeviceAssuranceIOS

File: okta/resource_okta_policy_device_assurance_macOS_test.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
+package okta
+
+import (
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func TestAccPolicyDeviceAssuranceMacOS(t *testing.T) {
```
Comment (monde) [inline okta/resource_okta_policy_device_assurance_macOS_test.go]: TestAccResourceOktaPolicyDeviceAssuranceMacOS

File: okta/resource_okta_policy_device_assurance_window_test.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
+package okta
+
+import (
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+// TODU add test for tpsp
+func TestAccPolicyDeviceAssuranceWindows(t *testing.T) {
```
Comment (monde) [inline okta/resource_okta_policy_device_assurance_window_test.go]: TestAccResourceOktaPolicyDeviceAssuranceWindows

**Comment by duytiennguyen-okta** [comment]
> @duytiennguyen-okta thank you so much for working on this. It sincerely means the world to me.

@lucascantor you mentioned in the issue(https://github.com/okta/terraform-provider-okta/issues/1240) that  ```okta_app_signon_policy_rule resources cannot [have Device Assurance policies added to them](https://help.okta.com/oie/en-us/Content/Topics/identity-engine/devices/device-assurance-policy-rule.htm) until an Okta Terraform provider version is released with support for this.```. What do you mean by that? Do you mean to create a new policy rule along with the device assurance? Can you give me an example script?

File: okta/resource_okta_policy_device_assurance_androi_test.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
+package okta
```
Comment (monde) [inline okta/resource_okta_policy_device_assurance_androi_test.go]: test file name has a typo ~~resource_okta_policy_device_assurance_androi_test.go~~ resource_okta_policy_device_assurance_android_test.go

File: okta/resource_okta_policy_device_assurance_androi.go [hunk 7]
```diff
@@ -0,0 +1,366 @@
+package okta
```
Comment (monde) [inline okta/resource_okta_policy_device_assurance_androi.go]: file name has a typo ~~resource_okta_policy_device_assurance_androi.go~~ resource_okta_policy_device_assurance_android.go

File: okta/resource_okta_policy_device_assurance_chromeOS.go [hunk 8]
```diff
@@ -0,0 +1,382 @@
+package okta
```
Comment (monde) [inline okta/resource_okta_policy_device_assurance_chromeOS.go]: Golang guideline is lower case file names ~~chromeOS.go~~ chromeos.go

File: okta/resource_okta_policy_device_assurance_chromeOS_test.go [hunk 9]
```diff
@@ -0,0 +1,61 @@
+package okta
```
Comment (monde) [inline okta/resource_okta_policy_device_assurance_chromeOS_test.go]: Golang guideline is lower case file names ~~chromeOS_test.go~~ chromeos_test.go

File: okta/resource_okta_policy_device_assurance_iOS.go [hunk 10]
```diff
@@ -0,0 +1,311 @@
+package okta
```
Comment (monde) [inline okta/resource_okta_policy_device_assurance_iOS.go]: Golang guideline is lower case file names ~~iOS.go~~ ios.go

File: okta/resource_okta_policy_device_assurance_iOS_test.go [hunk 11]
```diff
@@ -0,0 +1,45 @@
+package okta
```
Comment (monde) [inline okta/resource_okta_policy_device_assurance_iOS_test.go]: Golang guideline is lower case file names ~~iOS_test.go~~ ios_test.go

File: okta/resource_okta_policy_device_assurance_macOS.go [hunk 12]
```diff
@@ -0,0 +1,483 @@
+package okta
```
Comment (monde) [inline okta/resource_okta_policy_device_assurance_macOS.go]: Golang guideline is lower case file names ~~macOS.go~~ macos.go

File: okta/resource_okta_policy_device_assurance_macOS_test.go [hunk 13]
```diff
@@ -0,0 +1,77 @@
+package okta
```
Comment (monde) [inline okta/resource_okta_policy_device_assurance_macOS_test.go]: Golang guideline is lower case file names ~~macOS_test.go~~ macos_test.go

File: okta/resource_okta_policy_device_assurance_window.go [hunk 14]
```diff
@@ -0,0 +1,526 @@
+package okta
```
Comment (monde) [inline okta/resource_okta_policy_device_assurance_window.go]: Probably should be plural ~~window.go~~ windows.go

File: okta/resource_okta_policy_device_assurance_window_test.go [hunk 15]
```diff
@@ -0,0 +1,90 @@
+package okta
```
Comment (monde) [inline okta/resource_okta_policy_device_assurance_window_test.go]: Probably should be plural ~~window_test.go~~ windows_test.go

**Review by monde — CHANGES_REQUESTED** [review]

**Comment by lucascantor** [comment]
@duytiennguyen-okta to answer your question, when I made that comment on https://github.com/okta/terraform-provider-okta/issues/1240 I just meant that we can't yet perform the [`Add device assurance to an authentication policy`](https://help.okta.com/oie/en-us/Content/Topics/identity-engine/devices/device-assurance-policy-rule.htm) task via Okta's Terraform provider.

If it helps, here's an example of such an `okta_app_signon_policy_rule` resource, which has no Device Assurance policy added to it, because doing so is not yet supported by Okta's Terraform provider.

We're looking for a new (optional) argument to be added to the `okta_app_signon_policy_rule` resource, to allow us provide a list of Device Assurance policy ID(s) to be added to this Authentication Policy rule:

```hcl
resource "okta_app_signon_policy_rule" "critical_onboarded_users_managed_macos_windows" {
  policy_id = var.okta_auth_policies["critical"]
  name      = "Allow Onboarded Users - Managed w/ ZTA Score"
  priority  = 5
  access    = "ALLOW"
  groups_included = [
    okta_group.hr_inherited_role_groups["type_active_employees_not_onboarding"].id,
    okta_group.hr_inherited_role_groups["type_active_consultants_not_onboarding"].id,
  ]
  custom_expression           = var.crowdstrike_zta_auth_policy_rule_custom_expression
  device_is_managed           = true
  device_is_registered        = true
  factor_mode                 = "1FA"
  type                        = "ASSURANCE"
  re_authentication_frequency = "PT0S"
  constraints = [
    jsonencode({
      "possession" : {
        "types" : [
          "app"
        ],
        "methods" : [
          "signed_nonce"
        ],
        "hardwareProtection" : "REQUIRED",
        "deviceBound" : "REQUIRED",
        "userPresence" : "REQUIRED"
      }
    })
  ]
  depends_on = [okta_app_signon_policy_rule.critical_onboarded_users_managed_ios_android]
}
```

**Comment by duytiennguyen-okta** [comment]
@monde leave the file name be. I have issue with work space that require the file to be named that way, otherwise the workspace does not recognized the file

**Review by monde — APPROVED** [review]

**Pull request merged**